src/Interpreters/HashJoin.cpp

```diff
@@ -1498,31 +1907,61 @@ class NonJoinedBlockInputStream : private NotJoined, public IBlockInputStream
     template <ASTTableJoin::Strictness STRICTNESS, typename Map>
     size_t fillColumns(const Map & map, MutableColumns & columns_keys_and_right)
     {
-        using Mapped = typename Map::mapped_type;
-        using Iterator = typename Map::const_iterator;
-
         size_t rows_added = 0;
 
-        if (!position.has_value())
-            position = std::make_any<Iterator>(map.begin());
+        if constexpr (multiple_disjuncts)
+        {
+            if (!used_position.has_value())
+                used_position = parent.data->blocks.begin();
 
-        Iterator & it = std::any_cast<Iterator &>(position);
-        auto end = map.end();
+            auto end = parent.data->blocks.end();
 
-        for (; it != end; ++it)
+            for (auto & it = *used_position; it != end && rows_added < max_block_size; ++it)
+            {
+                const HashJoin::BlockWithFlags & block_with_flags = *it;
+
+                for (size_t row = 0; row < block_with_flags.flags.size(); ++row)
+                {
+                    if (!block_with_flags.flags[row])
+                    {
+                        for (size_t colnum = 0; colnum < columns_keys_and_right.size(); ++colnum)
+                        {
+                            auto clmn = block_with_flags.block.getByPosition(colnum).column;
+                            columns_keys_and_right[colnum]->insertFrom(*clmn, row);
```

Comment:
Yes, we do support ALL JON with OR.
The piece of code you've pointed out is about RIGHT or FULL join, we are adding "missed" rows.
It is based on another principles, we use flags (see BlockWithFlags) for every row.
I am not happy with coexisting these two approaches while I was worried about performance, that is why we have them both.

Reply:
Some more details.
The technique we have in master is based on HashMaps' offsetInternal. In our case HashMaps have different internal ordering and, more over, have different gaps because of NULLs. That is why a solution based on offsetInternal with reasonable calculation complexity seems hardly possible.
BlockWithFlags solves this problem keeping flags bound with data, not with a particular HashTable.